src/sentry/grouping/enhancer/__init__.py

```diff
@@ -474,6 +497,135 @@ def visit_quoted_ident(self, node, children):
         return node.match.groups()[0].lstrip("!")
 
 
+def _update_frames_from_cached_values(
+    frames: Sequence[dict[str, Any]], cache_key: str, platform: str, dry_run: bool = True
+) -> bool:
+    """
+    This will update the frames of the stacktrace if it's been cached.
+    Set dry_run to False to actually change the frames.
+    Returns if the merged has correctly happened.
+    """
+    frames_changed = False
+    # XXX: Test the fallback value
+    changed_frames_values = cache.get(cache_key, {})
+    # This helps tracking changes in the hit/miss ratio of the cache
+    metrics.incr(
+        "save_event.stacktrace.cache.get",
+        tags={"success": bool(changed_frames_values), "platform": platform},
+    )
+    if changed_frames_values:
+        try:
+            for frame, changed_frame_values in zip(frames, changed_frames_values):
+                if not dry_run:
+                    if changed_frame_values["in_app"] is not None:
+                        frame["in_app"] = changed_frame_values["in_app"]
+                        frames_changed = True
+                    if changed_frame_values["in_app"] is not None:
+                        set_path(frame, "data", "category", value=changed_frame_values["category"])
+                        frames_changed = True
+
+            if frames_changed:
+                logger.info("We have merged the cached stacktrace to the incoming one.")
+        except Exception as error:
+            logger.exception(
+                "We have failed to update the stacktrace from the cache. Not aborting execution.",
+                extra={"platform": platform},
+            )
+            # We want tests to fail to prevent breaking the caching system without noticing
+            if os.environ.get("PYTEST_CURRENT_TEST"):
+                raise error
+
+    metrics.incr(
+        "save_event.stacktrace.merged_cached_values",
+        tags={"success": frames_changed, "platform": platform},
+    )
+    return frames_changed
+
+
+def _cache_changed_frame_values(
+    frames: Sequence[dict[str, Any]], cache_key: str, platform: str
+) -> None:
+    """Store in the cache the values which have been modified for each frame."""
+    caching_succeeded = False
+    # Check that some other event has not already populated the cache
+    if cache.get(cache_key):
```

Comment:
This might be an expensive check, since we'll return the whole stack trace here. We could use `cache.touch()`  https://docs.djangoproject.com/en/4.2/topics/cache/#django.core.cache.cache.touch potentially, since it returns True if it successfully touches the key otherwise false. Might be worth reading the source to fully understand the behaviour though.

Reply:
We actually don't store the whole stacktrace but only the `in_app` and `category` values per frame.
I will add a metrics.timer to get a sense of the performance.

Is it with speed you're concerned about?

Source code to touch: [link to v3.2](https://github.com/django/django/blob/75418f8c0e5e8fcfb557bf8e69fc1ccfd64b697d/django/core/cache/backends/memcached.py#L86-L89)
I'm unclear if this would be what we want to do instead.